src/cargo/core/compiler/fingerprint.rs

```diff
@@ -1564,11 +1574,20 @@ pub fn translate_dep_info(
 
     let mut new_contents = Vec::new();
     for file in deps {
-        let file = rustc_cwd.join(file);
-        let (ty, path) = if let Ok(stripped) = file.strip_prefix(pkg_root) {
-            (DepInfoPathType::PackageRootRelative, stripped)
-        } else if let Ok(stripped) = file.strip_prefix(target_root) {
+        let file = if cfg!(windows) && file.starts_with("\\\\?\\") {
+            // Remove Windows extended-length prefix, since functions like
+            // strip_prefix won't work if you mix with traditional dos paths.
```

Comment:
Technically this isn't quite right I think because if you strip `\\?\C:\foo` from `\\?\C:\foo\bar` you'll get `bar`. If you strip `C:\foo` from `\\?\C:\foo\bar`, however, it doesn't work. I won't pretend to know enough about paths to know whether that's a bug or not though.

I think a better stategy may be to have canonicalized and non-canonicalized versions of the various root paths we're considering here? I think this would definitely be a problem on Unix as well if the project dir or something above it uses a symlink, but stripping both the current absolute dir as well as the canonicalized current absolute dir should solve that as well (and handle this `\\?\` business transparently)

Reply:
Sounds fair, I'll give it a shot.  It will just be a bit more complex, but hopefully not too much.  I'll also try to create a test to see if symlink canonicalization actually is a problem.

Reply:
Updated with a new approach which always compares canonical paths.

It did reveal that it was not handling symlinks properly, so it should work better on all platforms now.